#Overview of the Dataset
This dataset pertains to _eBay Kleinanzeigen_, a classifieds section from the German eBay website. Specifically, classifieds pertaining to used automobiles. The data is stored in the CSV file format.

#The Objective
The data provided needs to be cleaned. Once cleaned, an analysis of the data will performed using the NumPy and Pandas libraries.

In [1]:
import numpy as np
import pandas as pd


In [2]:
autos = pd.read_csv("autos.csv", encoding='Latin-1')
#Types of encoding: UTF-8, Windows-1252, Latin-1


In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [4]:
autos.info()
autos.head()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

#Observations Pertaining to the Data
Most of the columns have an object (string) data type, while some are int64 (integer). Looking at the price column, its data type is object. May want to convert the data type to integer or float. Same thing with the odometer column. 

yearOfRegistration, monthOfRegistration, and postalCode are all int64, it may make more sense to convert them to object, as they're not being used in numerical computations.

Lastly, may want to rename abtest to autoBrakeTest. It's unclear what the current name for the column is referring to.

In [5]:
autos.columns
        
        

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
autos.rename(columns={'dateCrawled':'date_crawled', 
                      'offerType':'offer_type',
                      'abtest':'autobrake_test',
                      'gearbox':'transmission',
                     'vehicleType':'vehicle_type',
                     'yearOfRegistration':'registration_year',
                     'powerPS':'horsepower',
                     'notRepairedDamage':'unrepaired_damage',
                    'monthOfRegistration':'registration_month',
                      'fuelType':'fuel_type',
                     'dateCreated':'creation_date',
                     'nrOfPictures':'total_pictures',
                     'postalCode':'postal_code',
                     'lastSeen':'last_seen'}, inplace=True)


In [7]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price autobrake_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000        control          bus               2004   
1  privat    Angebot  $8,500        control    limousine               1997   
2  privat    Angebot  $8,990           test    limousine               2009   
3  privat    Angebot  $4,350        control   kleinwagen               2007   
4  privat    Angebot  $1,350           test        kombi               2003   

  transmission  horsepower   model   odometer  registration_month fuel_type  \
0      manuell         158  andere  150,000km                   3       lpg   
1    automatik         286     7er  150,000km                   6    benzin   
2      manuell         102    golf   70,000km                   7    benzin   
3    automatik          71  fortwo   70,000km                   6    benzin   
4      manuell           0   focus  150,000km                   7    benzin   

        brand unrepaired_damage        creation_date  total_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

To meet with Python standards, I renamed the columns in camelcase to snakecase. Plus, there were a few columns with names that I found unclear as to what they meant. For example, I looked at the column titled 'gearbox', after observing the data it was referring to an automobile's transmission. Therefore, I renamed the column to 'transmission'.

In [8]:
autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine

In [9]:
autos.dtypes

date_crawled          object
name                  object
seller                object
offer_type            object
price                 object
autobrake_test        object
vehicle_type          object
registration_year      int64
transmission          object
horsepower             int64
model                 object
odometer              object
registration_month     int64
fuel_type             object
brand                 object
unrepaired_damage     object
creation_date         object
total_pictures         int64
postal_code            int64
last_seen             object
dtype: object

A couple of columns that might be worth dropping are 'seller' and 'total_pictures', as all the rows in both columns have the same values.

'unrepaired_damage', 'fuel_type', 'model', 'transmission', and 'vehicle_type' are worth investigating becuase there quite a few records with a value of NaN.

'registration_year', 'registration_month', and 'postal_code' should be converted to the object datatype because they're not used to make any calculations. Therefore, it doesn't make much sense to have the columns as int64. 'price' and 'odometer' should be converted to float.

Might consider changing the format of 'date_crawled' and 'last_seen' to 
YYY-MM-DD.

In [10]:
autos["price"] = autos["price"].str.replace("$", "")
autos["price"] = autos["price"].str.replace(",", "")
autos["odometer"] = autos["odometer"].str.replace("km", "")
autos["odometer"] = autos["odometer"].str.replace(",", "")

In [11]:
autos["price"] = autos["price"].astype(float)
autos["odometer"] = autos["odometer"].astype(float)

In [12]:
new_autos = autos.rename(columns={'odometer':'odometer_km'})

This new dataframe, new_autos, will take into account the changes made to the odometer and price columns.

In [13]:
new_autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine

In [14]:
new_autos["odometer_km"].shape

(50000,)

In [15]:
new_autos["price"].shape

(50000,)

In [16]:
new_autos["odometer_km"].unique().shape

(13,)

In [17]:
new_autos["price"].unique().shape

(2357,)

While both the odometer_km and price both have an equal number of values, the number of unique values for odometer_km differs drastically in comparison with price.

In [18]:
new_autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [19]:
new_autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

What caught my attention are the standard deviations for the respective columns. I am curious to see why the standard deviation for odometer_km is lower than the standard deviation for price, and what that says about the data.

According to the _National Library of Medicine_ (https://www.nlm.nih.gov/nichsr/stats_tutorial/section2/mod8_sd.html), "Low standard deviation means data are around the mean, and high standard deviation indicates data are more spread out." This can explain why odometer_km has very few unique values, in comparison to price. The standard deviation for odometer_km is low and most of its data is centered around the mean. 

With price, the data is spread out, and not clustered around the mean, that's because price has a much higher standard deviation than odometer_km. But, because price's data is more spread out, could there be the possibility of outliers in the data? Normally, outliers tend to be spread out and isolated from the other data points.

In [20]:
new_autos["odometer_km"].value_counts(ascending=True)

10000.0       264
20000.0       784
30000.0       789
40000.0       819
5000.0        967
50000.0      1027
60000.0      1164
70000.0      1230
80000.0      1436
90000.0      1757
100000.0     2169
125000.0     5170
150000.0    32424
Name: odometer_km, dtype: int64

In [21]:
new_autos["price"].value_counts().head()

0.0       1421
500.0      781
1500.0     734
2500.0     643
1200.0     639
Name: price, dtype: int64

value_counts() shows that the data for odometer_km is mostly centered around the mean. Very unlikely there's any outliers in that column. value_counts() for price, on the other hand, shows values varying drastically. This is a sign that there could be outliers in the price column.

In [22]:
no_outl_autos = new_autos[new_autos["price"].between(0,9000)]

In [23]:
no_outl_autos["price"].value_counts(ascending=True).head()

4349.0    1
4125.0    1
3965.0    1
4525.0    1
5180.0    1
Name: price, dtype: int64

I created another new dataframe to take into account the removal of the outliers in the price column. As shown, in the value_counts() for the dataframe, the data is less spread out.

In [24]:
no_outl_autos[['date_crawled', 'creation_date', 'last_seen']][0:5]

date_crawled        creation_date            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [25]:
print(no_outl_autos['date_crawled'].str[:10].head())

0    2016-03-26
1    2016-04-04
2    2016-03-26
3    2016-03-12
4    2016-04-01
Name: date_crawled, dtype: object


In [26]:
print(no_outl_autos['creation_date'].str[:10].head())

0    2016-03-26
1    2016-04-04
2    2016-03-26
3    2016-03-12
4    2016-04-01
Name: creation_date, dtype: object


In [27]:
print(no_outl_autos['last_seen'].str[:10].head())

0    2016-04-06
1    2016-04-06
2    2016-04-06
3    2016-03-15
4    2016-04-01
Name: last_seen, dtype: object


In [28]:
no_outl_autos["date_crawled"].value_counts(normalize=True, dropna=False).head()

2016-03-12 16:06:22    0.000074
2016-04-02 15:49:30    0.000074
2016-03-23 19:38:20    0.000074
2016-03-25 19:57:10    0.000074
2016-03-27 22:55:05    0.000074
Name: date_crawled, dtype: float64

In [29]:
no_outl_autos["creation_date"].value_counts(normalize=True, dropna=False).head()

2016-04-03 00:00:00    0.038296
2016-03-21 00:00:00    0.037730
2016-03-20 00:00:00    0.037705
2016-04-04 00:00:00    0.037041
2016-03-12 00:00:00    0.037016
Name: creation_date, dtype: float64

In [30]:
no_outl_autos["last_seen"].value_counts(normalize=True, dropna=False).head()

2016-04-06 15:16:45    0.000148
2016-04-06 07:46:11    0.000148
2016-04-07 06:17:27    0.000148
2016-04-07 02:45:51    0.000123
2016-03-15 14:44:36    0.000123
Name: last_seen, dtype: float64

In [31]:
no_outl_autos[['date_crawled', 'creation_date', 'last_seen']][0:5]

date_crawled        creation_date            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [35]:
print(no_outl_autos['date_crawled'].str[:10].head(5)) 
#This returns the first 10 characters of the date

0    2016-03-26
1    2016-04-04
2    2016-03-26
3    2016-03-12
4    2016-04-01
Name: date_crawled, dtype: object


In [36]:
print(no_outl_autos['creation_date'].str[:10].head(5))

0    2016-03-26
1    2016-04-04
2    2016-03-26
3    2016-03-12
4    2016-04-01
Name: creation_date, dtype: object


In [37]:
print(no_outl_autos['last_seen'].str[:10].head(5))

0    2016-04-06
1    2016-04-06
2    2016-04-06
3    2016-03-15
4    2016-04-01
Name: last_seen, dtype: object


In [50]:
no_outl_autos['creation_date'].value_counts(normalize=True, dropna=False).sort_index().head()

2015-08-10 00:00:00    0.000025
2015-11-10 00:00:00    0.000025
2015-12-05 00:00:00    0.000025
2015-12-30 00:00:00    0.000025
2016-01-07 00:00:00    0.000025
Name: creation_date, dtype: float64

In [51]:
no_outl_autos['date_crawled'].value_counts(normalize=True, dropna=False).sort_index().head()

2016-03-05 14:06:30    0.000025
2016-03-05 14:06:40    0.000025
2016-03-05 14:07:04    0.000025
2016-03-05 14:07:08    0.000025
2016-03-05 14:07:21    0.000025
Name: date_crawled, dtype: float64

In [52]:
no_outl_autos['last_seen'].value_counts(normalize=True, dropna=False).sort_index().head()

2016-03-05 14:45:46    0.000025
2016-03-05 14:49:34    0.000025
2016-03-05 15:16:11    0.000025
2016-03-05 15:16:47    0.000025
2016-03-05 15:28:10    0.000025
Name: last_seen, dtype: float64

The dates are formatted in YYYY-MM-DD with hours:minutes:seconds in military time. creation_date has data spanning from August 2015 to April 2016. While date_crawled and last_seen both have data spanning from March 2016 to April 2016.

In [53]:
no_outl_autos['registration_year'].min()

1000

In [54]:
no_outl_autos['registration_year'].max()

9999

The minimum for registration_year is 1000 and the maximum is 9999, both values are unrealistic. Let's assume the minimum is 1920, around the time a lot of people started to drive automobiles. The maximum will be set to 2021, the current year. I'll create a new dataframe that takes the changes in registration_year into account.

In [56]:
ry_autos = no_outl_autos[no_outl_autos["registration_year"].between(1920,2021)]

In [58]:
ry_autos['registration_year'].value_counts(normalize=True).head()

2000    0.081548
1999    0.073149
2005    0.066228
2001    0.065046
2003    0.063938
Name: registration_year, dtype: float64

The Registration Year with the highest percentage is 2000 (8.15%), second to that is 1999 (7.31%), and third is 2005 (6.62%). It can be assumed that a majority of the automobiles' Registration Year is 2000.

In [65]:
ry_autos['brand'].value_counts(normalize=True)

volkswagen        0.215630
opel              0.126570
bmw               0.092114
mercedes_benz     0.081203
ford              0.077385
audi              0.068075
renault           0.056943
peugeot           0.034112
fiat              0.030713
seat              0.020565
smart             0.017019
mazda             0.016576
citroen           0.015763
nissan            0.015319
skoda             0.014950
toyota            0.013374
hyundai           0.009901
volvo             0.009408
mitsubishi        0.009187
honda             0.008916
sonstige_autos    0.008596
alfa_romeo        0.007241
kia               0.007192
suzuki            0.006625
chevrolet         0.005369
mini              0.005345
chrysler          0.003990
daihatsu          0.003128
dacia             0.002512
subaru            0.002389
daewoo            0.001946
trabant           0.001872
saab              0.001773
rover             0.001675
jeep              0.001527
lancia            0.001281
jaguar            0.001108
l

In [76]:
ry_autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price autobrake_test vehicle_type  registration_year  \
0  privat    Angebot  5000.0        control          bus               2004   
1  privat    Angebot  8500.0        control    limousine               1997   
2  privat    Angebot  8990.0           test    limousine               2009   
3  privat    Angebot  4350.0        control   kleinwagen               2007   
4  privat    Angebot  1350.0           test        kombi               2003   

  transmission  horsepower   model  odometer_km  registration_month fuel_type  \
0      manuell         158  andere     150000.0                   3       lpg   
1    automatik         286     7er     150000.0                   6    benzin   
2      manuell         102    golf      70000.0                   7    benzin   
3    automatik          71  fortwo      70000.0                   6    benzin   
4      manuell           0   focus     150000.0                   7    benzin   

        brand unrepaired_damage        creation_date  total_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [77]:
top_brand_by_year = {}

auto_brands = ry_autos['brand'].unique()

In [84]:
for a in auto_brands:
    selected_rows = ry_autos[ry_autos['brand'] == a]
    sorted_rows = selected_rows.sort_values('date_crawled', ascending=False)
    brand = sorted_rows.iloc[7]['registration_year']
    top_brand_by_year[a]= brand


This for loop will aggregate data pertaining to the year and its top automobile brand. The rows are sorted in accordance with the date_crawled column.

In [92]:
top_brand_by_year

{'alfa_romeo': 1983,
 'audi': 1992,
 'bmw': 1997,
 'chevrolet': 2006,
 'chrysler': 1998,
 'citroen': 2004,
 'dacia': 2012,
 'daewoo': 2003,
 'daihatsu': 2002,
 'fiat': 2001,
 'ford': 2000,
 'honda': 2002,
 'hyundai': 2003,
 'jaguar': 1999,
 'jeep': 1994,
 'kia': 2012,
 'lada': 2002,
 'lancia': 2001,
 'land_rover': 2001,
 'mazda': 1999,
 'mercedes_benz': 1998,
 'mini': 2009,
 'mitsubishi': 2001,
 'nissan': 1997,
 'opel': 2010,
 'peugeot': 1999,
 'porsche': 1988,
 'renault': 2012,
 'rover': 2000,
 'saab': 1995,
 'seat': 2006,
 'skoda': 2005,
 'smart': 2010,
 'sonstige_autos': 2004,
 'subaru': 1995,
 'suzuki': 1995,
 'toyota': 2001,
 'trabant': 1983,
 'volkswagen': 2017,
 'volvo': 1999}